Scala compiler classpath gets confused by directory ending in .zip

we compile scala class files at runtime. the scala files we compile are configuration files in a custom dsl, mainly used for spark. for this we use the scala compiler in `scala.tools.nsc.Global` for scala 2.12.10

recently we started tested spark cluster deploy mode (so our code now runs on cluster in a jvm not directly created by us instead of in client jvm). as it turns out the jvm on the cluster had an unusual classpath including a directory named `__spark_conf__.zip`. this confused the classpath code used by the scala compiler, which tries to read this directory on the classpath as a file (presumably because its ending in .zip? maybe adding a simple File.isFile() check will solve this?)

below is a redacted stacktrace.
best,
koert

```
2020-05-26 13:14:47 ERROR yarn.ApplicationMaster: User class threw exception: scala.reflect.internal.FatalError: Error accessing /data/8/hadoop/nm/usercache/koert/filecache/722/__spark_conf__.zip
scala.reflect.internal.FatalError: Error accessing /data/8/hadoop/nm/usercache/koert/filecache/722/__spark_conf__.zip
	at scala.tools.nsc.classpath.AggregateClassPath.$anonfun$list$3(AggregateClassPath.scala:99)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.tools.nsc.classpath.AggregateClassPath.list(AggregateClassPath.scala:87)
	at scala.tools.nsc.util.ClassPath.list(ClassPath.scala:36)
	at scala.tools.nsc.util.ClassPath.list$(ClassPath.scala:36)
	at scala.tools.nsc.classpath.AggregateClassPath.list(AggregateClassPath.scala:30)
	at scala.tools.nsc.symtab.SymbolLoaders$PackageLoader.doComplete(SymbolLoaders.scala:284)
	at scala.tools.nsc.symtab.SymbolLoaders$SymbolLoader.complete(SymbolLoaders.scala:230)
	at scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1542)
	at scala.reflect.internal.Mirrors$RootsBase.init(Mirrors.scala:257)
	at scala.tools.nsc.Global.rootMirror$lzycompute(Global.scala:74)
	at scala.tools.nsc.Global.rootMirror(Global.scala:72)
	at scala.tools.nsc.Global.rootMirror(Global.scala:44)
	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass$lzycompute(Definitions.scala:295)
	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass(Definitions.scala:295)
	at scala.reflect.internal.Definitions$DefinitionsClass.init(Definitions.scala:1480)
	at scala.tools.nsc.Global$Run.<init>(Global.scala:1199)
        <redacted>
Caused by: java.io.IOException: Error accessing /data/8/hadoop/nm/usercache/koert/filecache/722/__spark_conf__.zip
	at scala.reflect.io.FileZipArchive.scala$reflect$io$FileZipArchive$$openZipFile(ZipArchive.scala:141)
	at scala.reflect.io.FileZipArchive.root$lzycompute(ZipArchive.scala:179)
	at scala.reflect.io.FileZipArchive.root(ZipArchive.scala:176)
	at scala.reflect.io.FileZipArchive.allDirs$lzycompute(ZipArchive.scala:212)
	at scala.reflect.io.FileZipArchive.allDirs(ZipArchive.scala:212)
	at scala.tools.nsc.classpath.ZipArchiveFileLookup.findDirEntry(ZipArchiveFileLookup.scala:76)
	at scala.tools.nsc.classpath.ZipArchiveFileLookup.list(ZipArchiveFileLookup.scala:63)
	at scala.tools.nsc.classpath.ZipArchiveFileLookup.list$(ZipArchiveFileLookup.scala:62)
	at scala.tools.nsc.classpath.ZipAndJarClassPathFactory$ZipArchiveClassPath.list(ZipAndJarFileLookupFactory.scala:56)
	at scala.tools.nsc.classpath.AggregateClassPath.$anonfun$list$3(AggregateClassPath.scala:91)
	... 35 more
Caused by: java.io.FileNotFoundException: /data/8/hadoop/nm/usercache/koert/filecache/722/__spark_conf__.zip (Is a directory)
	at java.util.zip.ZipFile.open(Native Method)
	at java.util.zip.ZipFile.<init>(ZipFile.java:225)
	at java.util.zip.ZipFile.<init>(ZipFile.java:155)
	at java.util.zip.ZipFile.<init>(ZipFile.java:169)
	at scala.reflect.io.FileZipArchive.scala$reflect$io$FileZipArchive$$openZipFile(ZipArchive.scala:138)
	... 44 more
```

--- a/src/compiler/scala/tools/nsc/classpath/FileUtils.scala
+++ b/src/compiler/scala/tools/nsc/classpath/FileUtils.scala
@@ -28,8 +28,7 @@ object FileUtils {
 
     def isScalaOrJavaSource: Boolean = !file.isDirectory && (file.hasExtension("scala") || file.hasExtension("java"))
 
-    // TODO do we need to check also other files using ZipMagicNumber like in scala.tools.nsc.io.Jar.isJarOrZip?
-    def isJarOrZip: Boolean = file.hasExtension("jar") || file.hasExtension("zip") || file.isInstanceOf[ZipArchive]
+    def isJarOrZip: Boolean = file.isInstanceOf[ZipArchive] || !file.isDirectory && (file.hasExtension("jar") || file.hasExtension("zip"))
 
     /**
      * Safe method returning a sequence containing one URL representing this file, when underlying file exists,

--- a/src/compiler/scala/tools/nsc/io/Jar.scala
+++ b/src/compiler/scala/tools/nsc/io/Jar.scala
@@ -169,11 +169,10 @@ object Jar {
   // See https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html
   // for some ideas.
   private val ZipMagicNumber = List[Byte](80, 75, 3, 4)
-  private def magicNumberIsZip(f: Path) = f.isFile && (f.toFile.bytes().take(4).toList == ZipMagicNumber)
+  private def magicNumberIsZip(f: Path) = f.toFile.bytes().take(4).toList == ZipMagicNumber
 
-  def isJarOrZip(f: Path): Boolean = isJarOrZip(f, examineFile = true)
-  def isJarOrZip(f: Path, examineFile: Boolean): Boolean =
-    f.hasExtension("zip", "jar") || (examineFile && magicNumberIsZip(f))
+  // file exists and either has name.jar or magic number
+  def isJarOrZip(f: Path): Boolean = f.isFile && (Path.isExtensionJarOrZip(f.name) || magicNumberIsZip(f))
 
   def create(file: File, sourceDir: Directory, mainClass: String): Unit = {
     val writer = new Jar(file).jarWriter(Name.MAIN_CLASS -> mainClass)

--- a/src/compiler/scala/tools/nsc/settings/MutableSettings.scala
+++ b/src/compiler/scala/tools/nsc/settings/MutableSettings.scala
@@ -16,7 +16,7 @@ package scala.tools
 package nsc
 package settings
 
-import io.{AbstractFile, Jar, Path, PlainFile, VirtualDirectory}
+import io.{AbstractFile, Path, PlainFile, VirtualDirectory}
 import scala.annotation.tailrec
 import scala.collection.mutable.Clearable
 import scala.io.Source
@@ -255,23 +255,29 @@ class MutableSettings(val errorFn: String => Unit, val pathFactory: PathFactory)
     /** Add a destination directory for sources found under `srcDir`.
      *  Both directories should exist.
      */
-    def add(srcDir: String, outDir: String): Unit = // used in ide?
-      add(checkDir(pathFactory.getDirectory(srcDir), srcDir),
-          checkDir(pathFactory.getDirectory(outDir), outDir))
+    // used in ide?
+    def add(srcDir: String, outDir: String): Unit = {
+      // Check that dir exists and is a directory.
+      def checkDir(name: String): AbstractFile = {
+        val dir = pathFactory.getDirectory(name)
+        if (dir != null && dir.isDirectory) dir
+        else throw new FatalError(s"$name does not exist or is not a directory")
+      }
+      add(checkDir(srcDir), checkDir(outDir))
+    }
 
-    /** Check that dir is exists and is a directory. */
-    private def checkDir(dir: AbstractFile, name: String, allowJar: Boolean = false): AbstractFile =
+    /** Check either existing dir, or if not dir in path, a jar/zip which may not yet exist. */
+    private def checkDirOrJar(name: String): AbstractFile = {
+      val dir = pathFactory.getDirectory(name)
       if (dir != null && dir.isDirectory) dir
-      else if (allowJar && dir == null && Jar.isJarOrZip(name, examineFile = false)) new PlainFile(Path(name))
+      else if (dir == null && Path.isExtensionJarOrZip(name)) new PlainFile(Path(name))
       else throw new FatalError(s"$name does not exist or is not a directory")
+    }
 
     /** Set the single output directory. From now on, all files will
      *  be dumped in there, regardless of previous calls to 'add'.
      */
-    def setSingleOutput(outDir: String): Unit = {
-      val dst = pathFactory.getDirectory(outDir)
-      setSingleOutput(checkDir(dst, outDir, allowJar = true))
-    }
+    def setSingleOutput(outDir: String): Unit = setSingleOutput(checkDirOrJar(outDir))
 
     def getSingleOutput: Option[AbstractFile] = singleOutDir
 

--- a/src/compiler/scala/tools/nsc/util/ClassPath.scala
+++ b/src/compiler/scala/tools/nsc/util/ClassPath.scala
@@ -19,7 +19,6 @@ import java.net.URL
 import java.util.regex.PatternSyntaxException
 
 import File.pathSeparator
-import Jar.isJarOrZip
 import scala.tools.nsc.classpath.{ClassPathEntries, PackageEntry, PackageName}
 
 /**
@@ -132,7 +131,7 @@ object ClassPath {
 
     /* Get all subdirectories, jars, zips out of a directory. */
     def lsDir(dir: Directory, filt: String => Boolean = _ => true) =
-      dir.list.filter(x => filt(x.name) && (x.isDirectory || isJarOrZip(x))).map(_.path).toList
+      dir.list.filter(x => filt(x.name) && (x.isDirectory || Jar.isJarOrZip(x))).map(_.path).toList
 
     if (pattern == "*") lsDir(Directory("."))
     else if (pattern endsWith wildSuffix) lsDir(Directory(pattern dropRight 2))

--- a/src/reflect/scala/reflect/io/AbstractFile.scala
+++ b/src/reflect/scala/reflect/io/AbstractFile.scala
@@ -48,7 +48,7 @@ object AbstractFile {
    */
   def getDirectory(file: File): AbstractFile =
     if (file.isDirectory) new PlainFile(file)
-    else if (file.isFile && Path.isExtensionJarOrZip(file.jfile)) ZipArchive fromFile file
+    else if (file.isFile && Path.isExtensionJarOrZip(file.jfile)) ZipArchive.fromFile(file)
     else null
 
   /**
@@ -63,7 +63,7 @@ object AbstractFile {
       else getFile(f)
     } else null
 
-  def getResources(url: URL): AbstractFile = ZipArchive fromManifestURL url
+  def getResources(url: URL): AbstractFile = ZipArchive.fromManifestURL(url)
 }
 
 /**

--- a/src/reflect/scala/reflect/io/Path.scala
+++ b/src/reflect/scala/reflect/io/Path.scala
@@ -36,9 +36,15 @@ import scala.util.Random.alphanumeric
  */
 object Path {
   def isExtensionJarOrZip(jfile: JFile): Boolean = isExtensionJarOrZip(jfile.getName)
-  def isExtensionJarOrZip(name: String): Boolean = {
-    name.endsWith(".jar") || name.endsWith(".zip")
-  }
+  def isExtensionJarOrZip(name: String): Boolean =
+    name.lastIndexOf('.') match {
+      case i if i >= 0 =>
+        val xt = name.substring(i + 1)
+        xt.equalsIgnoreCase("jar") || xt.equalsIgnoreCase("zip")
+      case _ => false
+    }
+
+  /** Lower case "extension", following the last dot. */
   def extension(name: String): String = {
     val i = name.lastIndexOf('.')
     if (i < 0) ""

--- a/src/testkit/scala/tools/testkit/TempDir.scala
+++ b/src/testkit/scala/tools/testkit/TempDir.scala
@@ -13,9 +13,11 @@
 package scala.tools.testkit
 
 import java.io.{IOException, File}
-import java.nio.file.{Files, Path}
+import java.nio.file.{Files, FileVisitor, FileVisitResult, SimpleFileVisitor, Path}, FileVisitResult.{CONTINUE => Continue}
+import java.nio.file.attribute._
 
 import scala.util.Properties
+import scala.util.Using.Releasable
 
 object TempDir {
   final val TEMP_DIR_ATTEMPTS = 10000
@@ -37,18 +39,28 @@ object TempDir {
  */
 case class ForDeletion(path: Path)
 object ForDeletion {
-  import scala.util.Using.Releasable
   implicit val deleteOnRelease: Releasable[ForDeletion] = new Releasable[ForDeletion] {
     override def release(releasee: ForDeletion) = if (!Properties.isWin) Files.delete(releasee.path)
   }
 }
 object ReleasablePath {
-  import scala.util.Using.Releasable
+  // On release of a path, delete the file it represents or recursively remove the directory.
   implicit val deleteOnRelease: Releasable[Path] = new Releasable[Path] {
-    override def release(releasee: Path) = if (!Properties.isWin) Files.delete(releasee)
+    override def release(releasee: Path) = if (!Properties.isWin) remove(releasee)
   }
+  // Delete a File on relese.
   implicit val deleteOnRelease2: Releasable[File] = new Releasable[File] {
     override def release(releasee: File) = if (!Properties.isWin) releasee.delete()
+  }
+
+  private def remove(path: Path): Unit = if (Files.isDirectory(path)) removeRecursively(path) else Files.delete(path)
+
+  private def removeRecursively(path: Path): Unit = Files.walkFileTree(path, new ZappingFileVisitor)
+
+  private class ZappingFileVisitor extends SimpleFileVisitor[Path] {
+    private def zap(path: Path) = { Files.delete(path) ; Continue }
+    override def postVisitDirectory(path: Path, e: IOException): FileVisitResult = if (e != null) throw e else zap(path)
+    override def visitFile(path: Path, attrs: BasicFileAttributes): FileVisitResult = zap(path)
   }
 }
 
@@ -56,7 +68,6 @@ object ReleasablePath {
  */
 object Releasables {
   import scala.reflect.io.ZipArchive
-  import scala.util.Using.Releasable
   implicit val closeZipOnRelease: Releasable[ZipArchive] = new Releasable[ZipArchive] {
     override def release(releasee: ZipArchive) = releasee.close()
   }

--- a/test/files/run/t12019/J_1.java
+++ b/test/files/run/t12019/J_1.java
@@ -1,0 +1,8 @@
+
+package p;
+
+public class J_1 {
+	public int f() {
+		return 42;
+	}
+}

--- a/test/files/run/t12019/Test.scala
+++ b/test/files/run/t12019/Test.scala
@@ -1,0 +1,32 @@
+
+import scala.tools.partest.DirectTest
+
+import scala.jdk.CollectionConverters._
+import scala.tools.testkit.ReleasablePath._
+import scala.util.Using
+
+import java.nio.file.Files._
+
+object Test extends DirectTest {
+
+  override def code: String = "class C { val c = new p.J_1().f() }"
+
+  override def show(): Unit =
+    Using.resource(createTempDirectory("t12019")) { dir =>
+      val target = createDirectory(dir.resolve("java.zip"))
+      val outdir = testOutput.jfile.toPath
+      val pkgdir = outdir.resolve("p")
+      Using.resource(walk(pkgdir)) { str =>
+        str.forEach { p =>
+          val partial = outdir.relativize(p)
+          val q = target.resolve(partial)
+          copy(p, q)
+        }
+      }
+      Using.resource(createTempDirectory("t12019out")) { out =>
+        val compiler = newCompiler(newSettings("-usejavacp" :: "-classpath" :: target.toString :: "-d" :: out.toString :: Nil))
+        compileString(compiler)(code)
+      }
+    }
+}
+// was: Error accessing /tmp/t120198214162953467729048/java.zip

--- a/test/files/run/t6240-universe-code-gen.scala
+++ b/test/files/run/t6240-universe-code-gen.scala
@@ -1,7 +1,6 @@
 
 import scala.jdk.CollectionConverters._
 import scala.tools.partest.nest.FileManager._
-import scala.util.Using
 
 import java.io.File
 import java.nio.file.Files

